File: src/main/java/de/danielmaile/aether/worldgen/AetherWorld.java
package de.danielmaile.aether.worldgen;

import de.danielmaile.aether.Aether;
import de.danielmaile.aether.util.Utils;
import org.bukkit.Bukkit;
import org.bukkit.World;

import javax.annotation.Nullable;
import java.io.*;
import java.net.URISyntaxException;

public class AetherWorld
{
    private static final String worldName = "world_aether_aether";

    public static ObjectManager getObjectManager()
    {
        return objectManager;
    }

    private static ObjectManager objectManager;

    @Nullable
    public static World getWorld()
    {
        return Bukkit.getWorld(worldName);
    }

    public static void init()
    {
        try
        {
            copyDatapack();
        }
        catch (IOException | URISyntaxException exception)
        {
            Aether.logError("Error while copying datapack. Please restart your server and try again!");
        }

        objectManager = new ObjectManager();
    }

    private static void copyDatapack() throws IOException, URISyntaxException
    {
        //Check if Datapack is already there
        String datapackPath = Bukkit.getWorldContainer() + File.separator + "world" + File.separator + "datapacks" + File.separator + "aether_datapack" + File.separator;
        if (new File(datapackPath).exists())
        {
            return;
        }

        //Copy Datapack
        saveResourceToWorldFolder("aether_datapack/data/aether/dimension/aether.json");
        saveResourceToWorldFolder("aether_datapack/data/aether/worldgen/biome/plains.json");
        saveResourceToWorldFolder("aether_datapack/pack.mcmeta");

        Bukkit.getConsoleSender().sendMessage(
                Aether.getLanguageManager().getComponent("messages.prefix")
                        .append(Aether.getLanguageManager().getComponent("messages.console.datapack_copied")));
    }

    private static void saveResourceToWorldFolder(String resourcePath) throws IOException
    {
        String outputPath = Bukkit.getWorldContainer() + File.separator + "world" + File.separator + "datapacks" + File.separator + resourcePath;
        InputStream inputStream = Utils.getResource(resourcePath);
        if (inputStream == null)
        {
            throw new IllegalArgumentException();
        }

        //Create output directory
        File outputDirectory = new File(outputPath.substring(0, outputPath.lastIndexOf('/')));
        if (!outputDirectory.exists())
        {
            if (!outputDirectory.mkdirs())
            {
                Aether.logError("Output directories for resource can't be created!");
                return;
            }
        }

        //Write file
        File outFile = new File(outputPath);
        OutputStream outputStream = new FileOutputStream(outFile);
        byte[] buf = new byte[1024];
        int length;
        while ((length = inputStream.read(buf)) > 0)
        {
            outputStream.write(buf, 0, length);
        }
        outputStream.close();
        inputStream.close();
    }
}

File: src/main/java/de/danielmaile/lama/aether/world/AetherWorld.kt
package de.danielmaile.lama.aether.world

import de.danielmaile.lama.aether.Aether.Companion.instance
import de.danielmaile.lama.aether.util.getResource
import de.danielmaile.lama.aether.util.logError
import org.bukkit.Bukkit
import org.bukkit.World
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream
import java.net.URISyntaxException

private const val worldName = "world_aether_aether"

class AetherWorld {

    val objectManager: ObjectManager = ObjectManager()

    init {
        try {
            copyDatapack()
        } catch (e: java.lang.Exception) {
            logError("Error while copying datapack. Please restart your server and try again!")
        }

        getWorld()?.populators?.add(BedrockPopulator())
    }

    fun getWorld(): World? {
        return Bukkit.getWorld(worldName)
    }

    @Throws(IOException::class, URISyntaxException::class)
    private fun copyDatapack() {
        //Check if Datapack is already there
        val datapackPath = Bukkit.getWorldContainer()
            .toString() + File.separator + "world" + File.separator + "datapacks" + File.separator + "aether_datapack" + File.separator
        if (File(datapackPath).exists()) {
            return
        }

        //Copy Datapack
        saveResourceToWorldFolder("aether_datapack/data/aether/dimension/aether.json")
        saveResourceToWorldFolder("aether_datapack/data/aether/worldgen/biome/plains.json")
        saveResourceToWorldFolder("aether_datapack/pack.mcmeta")

        Bukkit.getConsoleSender().sendMessage(
            instance.getLanguageManager().getComponent("messages.prefix")
                .append(instance.getLanguageManager().getComponent("messages.console.datapack_copied"))
        )
    }

    @Throws(IOException::class)
    private fun saveResourceToWorldFolder(resourcePath: String) {
        val outputPath = Bukkit.getWorldContainer()
            .toString() + File.separator + "world" + File.separator + "datapacks" + File.separator + resourcePath
        val inputStream = getResource(resourcePath) ?: throw IllegalArgumentException()

        //Create output directory
        val outputDirectory = File(outputPath.substring(0, outputPath.lastIndexOf('/')))
        if (!outputDirectory.exists() && !outputDirectory.mkdirs()) {
            logError("Output directories for resource can't be created!")
            return
        }

        //Write file
        val outFile = File(outputPath)
        val outputStream: OutputStream = FileOutputStream(outFile)
        val buf = ByteArray(1024)
        var length: Int
        while (inputStream.read(buf).also { length = it } > 0) {
            outputStream.write(buf, 0, length)
        }
        outputStream.close()
        inputStream.close()
    }
}